### PR TITLE
Support reinferring indexers

### PR DESCRIPTION
**Pull request by 333fred**
We now reinfer indexer nullability. Additionally, `GetSymbolInfo` was returning no information for range and pattern-based indexers, this has now been fixed to return the actual property or method symbol being used in those cases. @dotnet/roslyn-compiler for review. @jasonmalinowski and @ryzngard for the nullable api change, as well as for the general change for indexers and ranges. This might cause you to encounter symbols where you aren't currently.

### CODE REVIEW

File: src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs [hunk 1]
```diff
@@ -17,6 +20,23 @@ private CompilationVerifier CompileAndVerifyWithIndexAndRange(string s, string e
             return CompileAndVerify(comp, expectedOutput: expectedOutput);
         }
 
+        private (SemanticModel model, List<ElementAccessExpressionSyntax> accesses) GetModelAndAccesses(CSharpCompilation comp)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs]: >private [](start = 8, length = 7)

`static`

File: src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs [hunk 2]
```diff
@@ -17,6 +20,23 @@ private CompilationVerifier CompileAndVerifyWithIndexAndRange(string s, string e
...
+        {
+            var syntaxTree = comp.SyntaxTrees[0];
+            var root = syntaxTree.GetRoot();
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            return (model, root.DescendantNodes().OfType<ElementAccessExpressionSyntax>().ToList());
+        }
+
+        private void VerifyIndexCall(IMethodSymbol symbol, string methodName, string containingTypeName)
```
Comment (cston) [inline src/Compilers/CSharp/Test/Emit/CodeGen/IndexAndRangeTests.cs]: >private [](start = 8, length = 7)

`static`

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 3]
```diff
@@ -2719,5 +2719,167 @@ void verifyAnnotation(ExpressionSyntax expr, PublicNullableAnnotation annotation
...
+                // (18,16): warning CS8604: Possible null reference argument for parameter 'u' in 'object C<object, object>.this[object u]'.
+                //         _ = c3[o1];
+                Diagnostic(ErrorCode.WRN_NullReferenceArgument, "o1").WithArguments("u", "object C<object, object>.this[object u]").WithLocation(18, 16));
+
+            var syntaxTree = comp.SyntaxTrees[0];
+            var root = syntaxTree.GetRoot();
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var indexers = root.DescendantNodes().OfType<ElementAccessExpressionSyntax>().ToArray();
+            verifyAnnotation(indexers.AsSpan().Slice(0, 2), PublicNullableAnnotation.Annotated, PublicNullableAnnotation.NotAnnotated);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >.AsSpan() [](start = 37, length = 9)

Consider moving to `indexers` definition.

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 4]
```diff
@@ -2719,5 +2719,167 @@ void verifyAnnotation(ExpressionSyntax expr, PublicNullableAnnotation annotation
...
+
+            var comp = CreateCompilationWithIndexAndRangeAndSpan(source, options: WithNonNullTypesTrue());
+            comp.VerifyDiagnostics();
+
+            var syntaxTree = comp.SyntaxTrees[0];
+            var root = syntaxTree.GetRoot();
+            var model = comp.GetSemanticModel(syntaxTree);
+
+            var elementAccesses = root.DescendantNodes().OfType<ElementAccessExpressionSyntax>().ToArray();
+            verifyAnnotation(elementAccesses.AsSpan().Slice(0, 2), PublicNullableAnnotation.Annotated);
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >.AsSpan() [](start = 44, length = 9)

Move to `elementAccesses` definition?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs [hunk 5]
```diff
@@ -2719,5 +2719,167 @@ void verifyAnnotation(ExpressionSyntax expr, PublicNullableAnnotation annotation
...
+                void verifyIndexer(IPropertySymbol propertySymbol)
+                {
+                    Assert.True(propertySymbol.IsIndexer);
+                    Assert.Equal(annotation, propertySymbol.NullableAnnotation);
+                }
+            }
+        }
+
+        [Fact]
+        public void GetSymbolInfo_PatternReinferred()
```
Comment (cston) [inline src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: >Pattern [](start = 34, length = 7)

"Range"?

File: src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs (hunk 3 above)
Comment (RikkiGibson) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: is pattern index/range available here? :smile:

---
In reply to: [322903318](https://github.com/dotnet/roslyn/pull/38598#discussion_r322903318) [](ancestors = 322903318)

Comment (333fred) [reply src/Compilers/CSharp/Test/Symbol/Symbols/Source/NullablePublicAPITests.cs]: Nope, we don't have the types defined. @agocke was talking about maybe defining them though

---
In reply to: [322918319](https://github.com/dotnet/roslyn/pull/38598#discussion_r322918319) [](ancestors = 322918319,322903318)

**Comment by 333fred** [comment]
@cston addressed feedback.

**Pull request merged**